# Проект кинопроката
Таблица `mkrf_movies` содержит информацию из реестра прокатных удостоверений. У одного фильма может быть несколько прокатных удостоверений. 

- `title` — название фильма;
- `puNumber` — номер прокатного удостоверения;
- `show_start_date` — дата премьеры фильма;
- `type` — тип фильма;
- `film_studio` — студия-производитель;
- `production_country` — страна-производитель;
- `director` — режиссёр;
- `producer` — продюсер;
- `age_restriction` — возрастная категория;
- `refundable_support` *—* объём возвратных средств государственной поддержки;
- `nonrefundable_support` *—* объём невозвратных средств государственной поддержки;
- `financing_source` *—* источник государственного финансирования;
- `budget` *—* общий бюджет фильма;
- `ratings` *—* рейтинг фильма на КиноПоиске;
- `genres` *—* жанр фильма.

Обратите внимание, что столбец `budget` **уже включает в себя полный объём государственной поддержки. Данные в этом столбце указаны только для тех фильмов, которые получили государственную поддержку. 

Таблица `mkrf_shows` содержит сведения о показах фильмов в российских кинотеатрах.

- `puNumber` — номер прокатного удостоверения;
- `box_office` — сборы в рублях.

In [80]:
import os
import pandas as pd
import numpy as np
import re
import datetime
from fuzzywuzzy import process
from fuzzywuzzy import fuzz

pd.set_option("display.max_rows", None)
pd.set_option("display.max_columns", None)

## Базовые задачи

### Опишите проблемы типов имеющихся данных и варианты исправления

In [81]:
mkrf_movies = pd.read_csv('mkrf_movies.csv', dtype=str)
mkrf_movies.describe(include='all').T

count unique                              top  freq
title                  7486   6772          День святого Валентина_     5
puNumber               7486   7484                        221054410     2
show_start_date        7486   1965         2010-12-17T12:00:00.000Z    96
type                   7486      8                   Художественный  4520
film_studio            7468   5490            Киностудия "Мосфильм"   341
production_country     7484    950                              США  2172
director               7477   4811                       О.Семёнова    47
producer               6918   5349                        П.Смирнов    67
age_restriction        7486      5  «16+» - для детей старше 16 лет  2851
refundable_support      332     34                              0.0   221
nonrefundable_support   332     75                       25000000.0    53
budget                  332    290                              0.0    17
financing_source        332      3            Министерство культуры   164
ratings                6519     94                              6.6   312
genres                 6510    742                            драма   476

In [82]:
mkrf_shows = pd.read_csv('mkrf_shows.csv', dtype=str)
mkrf_shows.describe(include='all').T

count unique        top freq
puNumber    3158   3158  111000113    1
box_office  3158   2938        0.0   24

In [83]:
mkrf_merge = mkrf_movies.merge(mkrf_shows, how='left', on='puNumber').copy()
mkrf_merge.shape

(7486, 16)

In [84]:
mkrf_merge.head(2).T

0  \
title                                                   Открытый простор   
puNumber                                                       221048915   
show_start_date                                 2015-11-27T12:00:00.000Z   
type                                                      Художественный   
film_studio            Тачстоун Пикчерз, Кобальт Пикчерз, Бикон Пикче...   
production_country                                                   США   
director                                                   Кевин Костнер   
producer                       Дэвид Валдес, Кевин Костнер, Джейк Эбертс   
age_restriction                              «18+» - запрещено для детей   
refundable_support                                                   NaN   
nonrefundable_support                                                NaN   
budget                                                               NaN   
financing_source                                                     NaN   
ratings                                                              7.2   
genres                                            боевик,драма,мелодрама   
box_office                                                           NaN   

                                                   1  
title                           Особо важное задание  
puNumber                                   111013716  
show_start_date             2016-09-13T12:00:00.000Z  
type                                  Художественный  
film_studio                    Киностудия "Мосфильм"  
production_country                              СССР  
director                                   Е.Матвеев  
producer                                         NaN  
age_restriction        «6+» - для детей старше 6 лет  
refundable_support                               NaN  
nonrefundable_support                            NaN  
budget                                           NaN  
financing_source                                 NaN  
ratings                                          6.6  
genres                                 драма,военный  
box_office                                       NaN

In [85]:
mkrf_merge_backup = mkrf_merge.copy()
mkrf_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7486 entries, 0 to 7485
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   title                  7486 non-null   object
 1   puNumber               7486 non-null   object
 2   show_start_date        7486 non-null   object
 3   type                   7486 non-null   object
 4   film_studio            7468 non-null   object
 5   production_country     7484 non-null   object
 6   director               7477 non-null   object
 7   producer               6918 non-null   object
 8   age_restriction        7486 non-null   object
 9   refundable_support     332 non-null    object
 10  nonrefundable_support  332 non-null    object
 11  budget                 332 non-null    object
 12  financing_source       332 non-null    object
 13  ratings                6519 non-null   object
 14  genres                 6510 non-null   object
 15  box_office           

Видим много пропусков данных, но радует, что датафремы качественно сопоставились и все 3158 значений по кассовым сборам нашли свою пару.

### Болле подробно расммотрим все признаки:

1) `title` - кроме самого названия может содержать описание, год выпуска и альтернативное название. Решения:
* избавимся от описания, заключенного в символы `\` - отправим эти данные в отдельные признаки `first_lv_desc`, `second_lv_desc`.
* избавимся от прочей информации в скобках - поместим ее в `alter_title`

Обратим внимание, что год выхода фильма указанный в названии может отличаться от данных в `show_start_date` - стоит уточнить правила работы с признаками у заказчика.

In [86]:
mkrf_merge_get_decription = mkrf_merge['title'].str.rsplit("/",expand = True)
mkrf_merge_get_decription.columns = ['title_true', 'first_lv_desc', 'second_lv_desc']
mkrf_merge = pd.concat([mkrf_merge, mkrf_merge_get_decription], axis=1)

mkrf_merge['alter_title'] = mkrf_merge['title'].str.extract(r'(\(\w+\)$)')
mkrf_merge['alter_title'] = mkrf_merge['alter_title'].str.replace(r'\W','')
mkrf_merge['title_true'] = mkrf_merge['title_true'].str.replace(r'\([\w\W]+$','')

<ipython-input-86-03c2e6b20749>:6: FutureWarning: The default value of regex will change from True to False in a future version.
  mkrf_merge['alter_title'] = mkrf_merge['alter_title'].str.replace(r'\W','')
<ipython-input-86-03c2e6b20749>:7: FutureWarning: The default value of regex will change from True to False in a future version.
  mkrf_merge['title_true'] = mkrf_merge['title_true'].str.replace(r'\([\w\W]+$','')


Проверим результат, посмотрев на самые короткие названия - видим, что данное решение не всегда корректно:

In [87]:
# 7478
# 3705
mkrf_merge['title_true_len'] = mkrf_merge['title_true'].str.len()
mkrf_merge.sort_values('title_true_len')[['title', 'title_true', 'first_lv_desc',
 'second_lv_desc',
 'alter_title',
 'title_true_len',
 'type',
 'film_studio',
 'production_country']].head(10)

title title_true     first_lv_desc  second_lv_desc  \
977       (Не)жданный принц                         None            None   
7478  (Не)идеальный мужчина                         None            None   
5792                   9/11          9                11            None   
2241     З/Л/О: Новый вирус          З                 Л  О: Новый вирус   
1668                З/Л/О 2          З                 Л             О 2   
2242     З/Л/О: Новый вирус          З                 Л  О: Новый вирус   
3705    Не/смотря ни на что         Не  смотря ни на что            None   
3106                g (джи)         g               None            None   
1711                     ЖЖ         ЖЖ              None            None   
6825                     Мы         Мы              None            None   

     alter_title  title_true_len             type  \
977          NaN               0   Художественный   
7478         NaN               0   Художественный   
5792         NaN               1   Художественный   
2241         NaN               1   Художественный   
1668         NaN               1   Художественный   
2242         NaN               1   Художественный   
3705         NaN               2   Художественный   
3106         джи               2   Художественный   
1711         NaN               2   Художественный   
6825         NaN               2   Художественный   

                                            film_studio  \
977   Диджитал Фэктори, Европа Корп., ТФ 1 Фильм Про...   
7478                           ООО "Нон-Стоп Продакшн"    
5792  Блэк Беар Студиос, Филм Хауз, Спрокфеллер Пикч...   
2241       8383 Продакшнз, Блуди Дисгастинг, Коллектив    
1668  Блуди Дисгастинг, Коллектив, 8383 Продакшнз, С...   
2242       8383 Продакшнз, Блуди Дисгастинг, Коллектив    
3705          Сэвэн Пикчерс, СтудиоКанал, Зиглер Синема   
3106    Life is Short, Pin-Up Production, DPG, Отдел 13   
1711                                     ООО "ЛВ ФИЛЬМ"   
6825  Манкипо Продакшнз, Блюмхауз Продакшнз, Юниверс...   

            production_country  
977                    Франция  
7478                    Россия  
5792                       США  
2241                       США  
1668  США - Канада - Индонезия  
2242                       США  
3705                  Германия  
3106                    Россия  
1711                    Россия  
6825                       США

2) Очевидно что `puNumber` — номер прокатного удостоверения мы должны использовать как индексы. Изучим природу 2-х пар дублей и примем решение чуть позже.

In [88]:
mkrf_merge[mkrf_merge.duplicated(subset='puNumber', keep=False)]

title   puNumber  \
4638                   Как жениться и остаться холостым  221154310   
4639  Иоанна - женщина на папском престоле /По роман...  221154310   
5067                                    Анализируй это!  221054410   
5068                                     Анализируй то!  221054410   

               show_start_date            type  \
4638  2010-12-17T12:00:00.000Z  Художественный   
4639  2010-12-17T12:00:00.000Z  Художественный   
5067  2010-05-25T12:00:00.000Z  Художественный   
5068  2010-05-25T12:00:00.000Z  Художественный   

                                            film_studio  \
4638  Ше Вам, Скрипт Ассосье, Тэ Фэ 1 Фильм Продюксь...   
4639   Константин Фильм, А Эр Ди Дегето Фильм, Дюне ...   
5067          Уорнер Бразерс, Вилладж Роудшоу Филмз ЛТД   
5068  Уорнер Бразерс, Виллидж Роадшоу Пикчерз, Эн-Пи...   

                                production_country        director  \
4638                                       Франция    Эрик Лартиго   
4639  Германия - Великобритания - Италия - Испания  Зенке Вортманн   
5067                                 США-Австралия   ХЭрольд Рэмис   
5068                                           США  Гарольд Реймис   

                                        producer  \
4638                     Амандин Било, Алан Шаба   
4639  Оливер Бербен, Дорис Д.Хайнце, Фарук Элтан   
5067             Джейн Розенталь, Пола Уейнстейн   
5068             Джейн Розенталь, Паул Уэйнстейн   

                      age_restriction refundable_support  \
4638  «16+» - для детей старше 16 лет                NaN   
4639  «16+» - для детей старше 16 лет                NaN   
5067  «16+» - для детей старше 16 лет                NaN   
5068  «12+» - для детей старше 12 лет                NaN   

     nonrefundable_support budget financing_source ratings  \
4638                   NaN    NaN              NaN     7.0   
4639                   NaN    NaN              NaN     7.6   
5067                   NaN    NaN              NaN     7.4   
5068                   NaN    NaN              NaN     6.8   

                       genres box_office  \
4638        мелодрама,комедия        NaN   
4639  драма,мелодрама,история        NaN   
5067         комедия,криминал        NaN   
5068         комедия,криминал        NaN   

                                 title_true                   first_lv_desc  \
4638       Как жениться и остаться холостым                            None   
4639  Иоанна - женщина на папском престоле   По роману Донны Вулфольк Кросс   
5067                        Анализируй это!                            None   
5068                         Анализируй то!                            None   

     second_lv_desc alter_title  title_true_len  
4638           None         NaN              32  
4639                        NaN              37  
5067           None         NaN              15  
5068           None         NaN              14

3) `show_start_date` - переведем данные в тип данных 'дата' и отбросим время. Создадим столбws `show_start_year` и `show_start_month` для будущего анализа.

In [89]:
mkrf_merge['show_start_year'] = pd.to_datetime(mkrf_merge['show_start_date']).dt.tz_localize(None).dt.year
mkrf_merge['show_start_month'] = pd.to_datetime(mkrf_merge['show_start_date']).dt.tz_localize(None).dt. month
mkrf_merge['show_start_date'] = pd.to_datetime(mkrf_merge['show_start_date']).dt.tz_localize(None).dt.date

4) Для убедимся в сопоставимости и качестве значений для признаков `type`, `age_restriction`, `financing_source` — структурирем категории, где это необходимо. Создадим числовой признак `age_limit`, на базе строкового `age_restriction`.

In [90]:
mkrf_merge['type'] = mkrf_merge['type'].str.lstrip()
mkrf_merge['type'] = mkrf_merge['type'].str.rstrip()

In [91]:
mkrf_merge['age_limit'] = mkrf_merge['age_restriction'].str[1:3]
mkrf_merge['age_limit'] = mkrf_merge['age_limit'].str.replace('\+','')
mkrf_merge['age_limit'].value_counts()

<ipython-input-91-1b9d2f96d961>:2: FutureWarning: The default value of regex will change from True to False in a future version.
  mkrf_merge['age_limit'] = mkrf_merge['age_limit'].str.replace('\+','')


16    2851
18    1605
12    1593
0      811
6      626
Name: age_limit, dtype: int64

5) Посмотрим на значения в `ratings`. Изучим фильмы которые оценены процентно на сайте КиноПоиска - обнаружить какую-то логику не получилось. Создадим для этих фильмов отдельный набор и позже пример решение, как работать с этими данными.

In [92]:
mkrf_merge[(~mkrf_merge['ratings'].isna()) \
           & (mkrf_merge['ratings'].str.contains('%'))].sort_values(by='ratings')[[
'title_true',
'director','producer',                                                                                                    
'title',
'type',
'ratings',
'genres',
'box_office',
'alter_title',
'show_start_year',
'show_start_month']].sample(7)

title_true  \
5455                                   Вечный холод   
3139                                  Я умею вязать   
6426  Памятные даты России. Партизаны и подпольщики   
259                                         Прошлое   
6737                                 Колесо времени   
5910                                           Мира   
5332                                        Добыча.   

                               director  \
5455                          А.Мигачев   
3139  Н.Степанова при участии С.Иванова   
6426             Н.Иванова, А.Кузнецова   
259                       Эктор Бабенко   
6737                         Б.Куломзин   
5910                           Д.Шабаев   
5332                    Патрик Сиверсен   

                                      producer  \
5455                                 И.Бачурин   
3139                 С.Кикнавелидзе, Д.Улюкаев   
6426                               В.Коханович   
259   Эктор Бабенко, Оскар Крамер, Хуго Сидмэн   
6737                                Б.Куломзин   
5910                               Н.Мокрицкая   
5332               Закари Ти Брайан, Кристофер   

                                              title            type ratings  \
5455                                   Вечный холод  Художественный     83%   
3139                                  Я умею вязать  Художественный     90%   
6426  Памятные даты России. Партизаны и подпольщики  Документальный     91%   
259                                         Прошлое  Художественный     99%   
6737                                 Колесо времени  Художественный     98%   
5910                                           Мира  Документальный     94%   
5332                                        Добыча.  Художественный     88%   

                         genres box_office alter_title  show_start_year  \
5455     фэнтези,боевик,триллер        NaN         NaN             2017   
3139                      драма   801370.0         NaN             2015   
6426            драма,биография        NaN         NaN             2018   
259         ужасы,триллер,драма        NaN         NaN             2015   
6737        фэнтези,приключения    34992.0         NaN             2019   
5910                      драма        NaN         NaN             2018   
5332  фантастика,боевик,триллер        NaN         NaN             2011   

      show_start_month  
5455                 8  
3139                11  
6426                11  
259                  7  
6737                 3  
5910                 6  
5332                 6

In [93]:
mkrf_no_rating_backup = mkrf_merge[(~mkrf_merge['ratings'].isna()) \
           & (mkrf_merge['ratings'].str.contains('%'))]

mkrf_no_rating_backup.shape

(29, 24)

6)  Далее нам надо привести следующие столбцы к числовому формату:

* 'refundable_support',
* 'nonrefundable_support',
* 'budget',
* 'ratings',
* 'box_office',
* age_limit

In [94]:
for col in ['refundable_support',
 'nonrefundable_support',
 'budget',
 'ratings',
 'box_office',
'age_limit']:
    mkrf_merge[col] = mkrf_merge[col].str.rstrip()
    mkrf_merge[col] = mkrf_merge[col].str.lstrip()
    mkrf_merge[col] = pd.to_numeric(mkrf_merge[col], errors='coerce')

In [95]:
mkrf_merge.sample(5).T

6120  \
title                                                         Слендермен   
puNumber                                                       121017818   
show_start_date                                               2018-08-16   
type                                                      Художественный   
film_studio               Митолоджи Интертейнмент, Мэдхаус Интертейнмент   
production_country                                                   США   
director                                                    Силвейн Уайт   
producer               Джеймс Вандербилт, Робин Мейсингер, Сара Сноу,...   
age_restriction                          «16+» - для детей старше 16 лет   
refundable_support                                                   NaN   
nonrefundable_support                                                NaN   
budget                                                               NaN   
financing_source                                                     NaN   
ratings                                                              4.2   
genres                                                    ужасы,детектив   
box_office                                                   95167526.89   
title_true                                                    Слендермен   
first_lv_desc                                                       None   
second_lv_desc                                                      None   
alter_title                                                          NaN   
title_true_len                                                        10   
show_start_year                                                     2018   
show_start_month                                                       8   
age_limit                                                             16   

                                                                    998   \
title                  Призрачный патруль /По мотивам одноименного ко...   
puNumber                                                       121012513   
show_start_date                                               2013-05-07   
type                                                      Художественный   
film_studio            Дак Хос Интертейнмент, Ориджинал Филм, Эр Эс И...   
production_country                                                   США   
director                                                  Роберт Швентке   
producer                      Нил Х.Мориц, Майк Ричардсон, Майкл Фоттрел   
age_restriction                          «12+» - для детей старше 12 лет   
refundable_support                                                   NaN   
nonrefundable_support                                                NaN   
budget                                                               NaN   
financing_source                                                     NaN   
ratings                                                              6.3   
genres                                            фэнтези,боевик,комедия   
box_office                                                           NaN   
title_true                                           Призрачный патруль    
first_lv_desc           По мотивам одноименного комикса Питера М.Ленкова   
second_lv_desc                                                             
alter_title                                                          NaN   
title_true_len                                                        19   
show_start_year                                                     2013   
show_start_month                                                       5   
age_limit                                                             12   

                                                                  6057  \
title                                                Проигранное место   
puNumber                                                     111012218   
show_start_date                             

7) Так же мы видим, что над фильмом могут работать большие коллективы: несколько студий, стран, режисеров и продюссеров. Весьма вероятно, что это важный фактор при оценке сложности работы над фильмом. Произведем подсчет участников для каждого фильма в признаках
* 'film_studio',
* 'production_country',
* 'director',
* 'producer'

Обратим внимание, что перечисление везде ведется через запятую, но в случае со странами, запятую иногда заменяют тире.

In [96]:
for col in ['film_studio', 'production_country', 'director', 'producer']:
    mkrf_merge[f'counts_'+col] = mkrf_merge[col].str.count(',')
    
    
mkrf_merge[f'counts_production_country'] += mkrf_merge['production_country'].str.count('-')

In [97]:
mkrf_merge.describe(include='all').T

count unique                              top  \
title                        7486   6772          День святого Валентина_   
puNumber                     7486   7484                        221054410   
show_start_date              7486   1965                       2010-12-17   
type                         7486      6                   Художественный   
film_studio                  7468   5490            Киностудия "Мосфильм"   
production_country           7484    950                              США   
director                     7477   4811                       О.Семёнова   
producer                     6918   5349                        П.Смирнов   
age_restriction              7486      5  «16+» - для детей старше 16 лет   
refundable_support          332.0    NaN                              NaN   
nonrefundable_support       332.0    NaN                              NaN   
budget                      332.0    NaN                              NaN   
financing_source              332      3            Министерство культуры   
ratings                    6490.0    NaN                              NaN   
genres                       6510    742                            драма   
box_office                 3158.0    NaN                              NaN   
title_true                   7486   6716                         Фиксики    
first_lv_desc                 519    365            По роману Дж.К.Ролинг   
second_lv_desc                490     53                                    
alter_title                   187     37                             2016   
title_true_len             7486.0    NaN                              NaN   
show_start_year            7486.0    NaN                              NaN   
show_start_month           7486.0    NaN                              NaN   
age_limit                  7486.0    NaN                              NaN   
counts_film_studio         7468.0    NaN                              NaN   
counts_production_country  7484.0    NaN                              NaN   
counts_director            7477.0    NaN                              NaN   
counts_producer            6918.0    NaN                              NaN   

                           freq              mean               std     min  \
title                         5               NaN               NaN     NaN   
puNumber                      2               NaN               NaN     NaN   
show_start_date              96               NaN               NaN     NaN   
type                       5909               NaN               NaN     NaN   
film_studio                 341               NaN               NaN     NaN   
production_country         2172               NaN               NaN     NaN   
director                     47               NaN               NaN     NaN   
producer                     67               NaN               NaN     NaN   
age_restriction            2851               NaN               NaN     NaN   
refundable_support          NaN   11864457.831325   24916555.263644     0.0   
nonrefundable_support       NaN   48980988.891566   59980117.923301     0.0   
budget                      NaN  127229716.683735  188588333.115158     0.0   
financing_source            164               NaN               NaN     NaN   
ratings                     NaN          6.479245          1.101665     1.0   
genres                      476               NaN               NaN     NaN   
box_office                  NaN   76478696.161219   240353122.82324     0.0   
title_true                    8               NaN               NaN     NaN   
first_lv_desc                21               NaN               NaN     NaN   
second_lv_desc              392               NaN               NaN     NaN   
alter_title                  61               NaN               NaN     NaN   
title_true_len              NaN         17.696634         12.042455     0.0   
show_start_year             NaN       2014.594176       

In [98]:
mkrf_merge[mkrf_merge['counts_production_country'] == 9]['production_country'][1724]

'Мексика - Эквадор - Канада - США - Франция - Малайзия - Италия - Аргентина - Германия - Индия  '

In [99]:
mkrf_merge[mkrf_merge['counts_producer'] == 14]['producer'][1724]

"Натали Галазка, Кристиан Гемеинер, Родриго Доносо,  Милко Коронед, Аншут Кумар, Ваи Ленг, Марк Никелсбург, Мелисса С.О'Брайн, Флориан Сиегл, Стефан Сках, Джеймс Стюарт, Доменико Тискони, Чеонг, Федерико Эйер, Джереми Ясхес"

In [188]:
w_df = mkrf_merge.copy()

### Исследуйте дубликаты и пропуски. Какова их природа в этом наборе данных? Можно ли это исправить?

Смотрим на дубли по признаку `puNumber`:

Очевидный дубль `Анализируй это!` - просто проверим в [источнике](https://www.kinopoisk.ru/film/1594/)

In [189]:
w_df[w_df.duplicated(subset='puNumber', keep=False)]

title   puNumber  \
4638                   Как жениться и остаться холостым  221154310   
4639  Иоанна - женщина на папском престоле /По роман...  221154310   
5067                                    Анализируй это!  221054410   
5068                                     Анализируй то!  221054410   

     show_start_date            type  \
4638      2010-12-17  Художественный   
4639      2010-12-17  Художественный   
5067      2010-05-25  Художественный   
5068      2010-05-25  Художественный   

                                            film_studio  \
4638  Ше Вам, Скрипт Ассосье, Тэ Фэ 1 Фильм Продюксь...   
4639   Константин Фильм, А Эр Ди Дегето Фильм, Дюне ...   
5067          Уорнер Бразерс, Вилладж Роудшоу Филмз ЛТД   
5068  Уорнер Бразерс, Виллидж Роадшоу Пикчерз, Эн-Пи...   

                                production_country        director  \
4638                                       Франция    Эрик Лартиго   
4639  Германия - Великобритания - Италия - Испания  Зенке Вортманн   
5067                                 США-Австралия   ХЭрольд Рэмис   
5068                                           США  Гарольд Реймис   

                                        producer  \
4638                     Амандин Било, Алан Шаба   
4639  Оливер Бербен, Дорис Д.Хайнце, Фарук Элтан   
5067             Джейн Розенталь, Пола Уейнстейн   
5068             Джейн Розенталь, Паул Уэйнстейн   

                      age_restriction  refundable_support  \
4638  «16+» - для детей старше 16 лет                 NaN   
4639  «16+» - для детей старше 16 лет                 NaN   
5067  «16+» - для детей старше 16 лет                 NaN   
5068  «12+» - для детей старше 12 лет                 NaN   

      nonrefundable_support  budget financing_source  ratings  \
4638                    NaN     NaN              NaN      7.0   
4639                    NaN     NaN              NaN      7.6   
5067                    NaN     NaN              NaN      7.4   
5068                    NaN     NaN              NaN      6.8   

                       genres  box_office  \
4638        мелодрама,комедия         NaN   
4639  драма,мелодрама,история         NaN   
5067         комедия,криминал         NaN   
5068         комедия,криминал         NaN   

                                 title_true                   first_lv_desc  \
4638       Как жениться и остаться холостым                            None   
4639  Иоанна - женщина на папском престоле   По роману Донны Вулфольк Кросс   
5067                        Анализируй это!                            None   
5068                         Анализируй то!                            None   

     second_lv_desc alter_title  title_true_len  show_start_year  \
4638           None         NaN              32             2010   
4639                        NaN              37             2010   
5067           None         NaN              15             2010   
5068           None         NaN              14             2010   

      show_start_month  age_limit  counts_film_studio  \
4638                12         16                 4.0   
4639                12         16                 6.0   
5067                 5         16                 1.0   
5068                 5         12                 4.0   

      counts_production_country  counts_director  counts_producer  
4638                        0.0              0.0              1.0  
4639                        3.0              0.0              2.0  
5067                        1.0              0.0              1.0  
5068                        0.0              0.0              1.0

In [190]:
# удалим строку с индексом 5068
w_df = w_df.drop(5068)
w_df[w_df.duplicated(subset='puNumber', keep=False)]

title   puNumber  \
4638                   Как жениться и остаться холостым  221154310   
4639  Иоанна - женщина на папском престоле /По роман...  221154310   

     show_start_date            type  \
4638      2010-12-17  Художественный   
4639      2010-12-17  Художественный   

                                            film_studio  \
4638  Ше Вам, Скрипт Ассосье, Тэ Фэ 1 Фильм Продюксь...   
4639   Константин Фильм, А Эр Ди Дегето Фильм, Дюне ...   

                                production_country        director  \
4638                                       Франция    Эрик Лартиго   
4639  Германия - Великобритания - Италия - Испания  Зенке Вортманн   

                                        producer  \
4638                     Амандин Било, Алан Шаба   
4639  Оливер Бербен, Дорис Д.Хайнце, Фарук Элтан   

                      age_restriction  refundable_support  \
4638  «16+» - для детей старше 16 лет                 NaN   
4639  «16+» - для детей старше 16 лет                 NaN   

      nonrefundable_support  budget financing_source  ratings  \
4638                    NaN     NaN              NaN      7.0   
4639                    NaN     NaN              NaN      7.6   

                       genres  box_office  \
4638        мелодрама,комедия         NaN   
4639  драма,мелодрама,история         NaN   

                                 title_true                   first_lv_desc  \
4638       Как жениться и остаться холостым                            None   
4639  Иоанна - женщина на папском престоле   По роману Донны Вулфольк Кросс   

     second_lv_desc alter_title  title_true_len  show_start_year  \
4638           None         NaN              32             2010   
4639                        NaN              37             2010   

      show_start_month  age_limit  counts_film_studio  \
4638                12         16                 4.0   
4639                12         16                 6.0   

      counts_production_country  counts_director  counts_producer  
4638                        0.0              0.0              1.0  
4639                        3.0              0.0              2.0

Другая пара явно разные фильмы, мы не хотим пока терять данные, поэтому просто добавим дополнительный символ в один из номеров прокатного удостоверения (предварительно убедимя, что номер станет уникальным), но запмним это на будующее - убедимся в правомерности решения позже, при общении с заказчиком.

In [191]:
w_df.loc[4639,'puNumber'] = '2211543102'

In [192]:
# пробуем создать числовые индексы
w_df['puNumber'] = pd.to_numeric(w_df['puNumber'], errors='raise')

ValueError: Unable to parse string "нет" at position 1797

Видим еще одно странное значение, так же исправим произвольным номером.

In [193]:
w_df[w_df['puNumber'] == '0']

Empty DataFrame
Columns: [title, puNumber, show_start_date, type, film_studio, production_country, director, producer, age_restriction, refundable_support, nonrefundable_support, budget, financing_source, ratings, genres, box_office, title_true, first_lv_desc, second_lv_desc, alter_title, title_true_len, show_start_year, show_start_month, age_limit, counts_film_studio, counts_production_country, counts_director, counts_producer]
Index: []

In [194]:
w_df.loc[1797,'puNumber'] = '0'

In [195]:
w_df['puNumber'] = pd.to_numeric(w_df['puNumber'], errors='raise')

In [196]:
w_df = w_df.set_index('puNumber')
w_df.sample(3).T

puNumber                                     121016511  \
title                               Смертельный звонок   
show_start_date                             2011-09-09   
type                                    Художественный   
film_studio                         Кор Контентс Медиа   
production_country                         Южная Корея   
director                                    Ю Сун-Донг   
producer                                Хонг Джунг-пьё   
age_restriction            «18+» - запрещено для детей   
refundable_support                                 NaN   
nonrefundable_support                              NaN   
budget                                             NaN   
financing_source                                   NaN   
ratings                                            4.9   
genres                                           ужасы   
box_office                                         NaN   
title_true                          Смертельный звонок   
first_lv_desc                                     None   
second_lv_desc                                    None   
alter_title                                        NaN   
title_true_len                                      18   
show_start_year                                   2011   
show_start_month                                     9   
age_limit                                           18   
counts_film_studio                                 0.0   
counts_production_country                          0.0   
counts_director                                    0.0   
counts_producer                                    0.0   

puNumber                                                           221083610  \
title                                                              Суррогаты   
show_start_date                                                   2010-07-15   
type                                                          Художественный   
film_studio                Мэндевиль Филмз, Тачстоун Пикчерз, Роуд Ребел,...   
production_country                                                       США   
director                                                     Джонатан Мостоу   
producer                       Тодд Либерманн, Дэвид Хоберман, Макс Хэндлман   
age_restriction                              «16+» - для детей старше 16 лет   
refundable_support                                                       NaN   
nonrefundable_support                                                    NaN   
budget                                                                   NaN   
financing_source                                                         NaN   
ratings                                                                  6.9   
genres                                             фантастика,боевик,триллер   
box_office                                                               NaN   
title_true                                                         Суррогаты   
first_lv_desc                                                           None   
second_lv_desc                                                          None   
alter_title                                                              NaN   
title_true_len                                                             9   
show_start_year                                                         2010   
show_start_month                                                           7   
age_limit                                                                 16   
counts_film_studio                                                       3.0   
counts_production_country                                                0.0   
counts_director                                                          0.0   
counts_producer                                                          2.0   

puNumber                                              112004319  
title                                              Поэт и война  
sh

### Датасет содержит много пропусков, и необходимо сделать выборку важных для нас признаков, посмотреть на объем оставшихся данных. На мой взгляд оптимальным решением будет дополнительное обогащение данных из внешних источников:

1) [Данные о рейтингах, создателях, описании фильма](https://www.kinopoisk.ru/)

2) [Данные о кассовых сборах](https://opendata.mkrf.ru/opendata/7705851331-movie_gross)

3) [Данные о гос. инвестировании](https://ekinobilet.fond-kino.ru/government-support/)

### Посчитайте динамику сборов по годам премьер фильмов. В каком году медианная сумма сборов была максимальной?

In [197]:
box_median_dict = {}

for year in sorted(w_df['show_start_year'].unique().tolist()):
    box_median = w_df[w_df['show_start_year'] == year]['box_office'].median()
    box_median_dict[year] = box_median
    print(f'{year} медианная сумма сборов: {box_median}\n')
    
print('Максимальная медиианная сумма сборов была в', max(box_median_dict, key=box_median_dict.get), ' году.')

2010 медианная сумма сборов: 1700.0

2011 медианная сумма сборов: 3000.0

2012 медианная сумма сборов: 5660.0

2013 медианная сумма сборов: 3522.5

2014 медианная сумма сборов: 18160.0

2015 медианная сумма сборов: 4920933.0

2016 медианная сумма сборов: 3846679.41

2017 медианная сумма сборов: 9968340.0

2018 медианная сумма сборов: 8891102.21

2019 медианная сумма сборов: 4627798.345

Максимальная медиианная сумма сборов была в 2017  году.


### С помощью сводной таблицы посчитайте среднюю и медианную сумму сборов для каждого года.

In [198]:
w_df['box_office_mean'] = w_df['box_office']
w_df['box_office_median'] = w_df['box_office']

pivot_mean_and_median_box = pd.pivot_table(w_df, values=['box_office_mean',
                                                         'box_office_median'], index=['show_start_year'],
                    aggfunc={'box_office_mean': np.mean,
                             'box_office_median': np.median})

pivot_mean_and_median_box.astype('int64')

box_office_mean  box_office_median
show_start_year                                    
2010                       23130               1700
2011                      129383               3000
2012                       54339               5660
2013                      161954               3522
2014                    26684415              18160
2015                    84940570            4920933
2016                    91000569            3846679
2017                   136032793            9968340
2018                   104565059            8891102
2019                    91369261            4627798

### Выделите имена продюсеров из столбца producer. У кого из них самый высокий средний рейтинг фильмов?

In [199]:
producers_average_rating = w_df[['title_true','producer','ratings']]
producers_average_rating.head(2)

title_true                                   producer  \
puNumber                                                                     
221048915      Открытый простор  Дэвид Валдес, Кевин Костнер, Джейк Эбертс   
111013716  Особо важное задание                                        NaN   

           ratings  
puNumber            
221048915      7.2  
111013716      6.6

In [200]:
producers_average_rating.shape

(7485, 3)

In [201]:
producers_average_rating = producers_average_rating.dropna()
producers_average_rating.shape

(6011, 3)

In [202]:
producers_average_rating['producer'] = producers_average_rating['producer'].str.split(',')
producers_average_rating.head(2)

title_true  \
puNumber                      
221048915  Открытый простор   
221038416      Особо опасен   

                                                    producer  ratings  
puNumber                                                               
221048915      [Дэвид Валдес,  Кевин Костнер,  Джейк Эбертс]      7.2  
221038416  [Джим Лемли,  Джейсон Нетер,  Марк Е.Платт,  Я...      6.8

In [203]:
producers_average_rating_explode = producers_average_rating.explode('producer').drop_duplicates()
producers_average_rating_explode = producers_average_rating_explode.reset_index()
producers_average_rating_explode.shape

(13156, 4)

In [204]:
producers_average_rating_explode.head(2)

puNumber        title_true        producer  ratings
0  221048915  Открытый простор    Дэвид Валдес      7.2
1  221048915  Открытый простор   Кевин Костнер      7.2

In [205]:
pivot_producers_average_rating = pd.pivot_table(producers_average_rating_explode, values='ratings',
                                                                                     index=['producer'],
                    aggfunc={'ratings': np.mean})


pivot_producers_average_rating.sort_values(by='ratings', ascending=False).head(10)

ratings
producer                         
 Ники Марвин                  9.1
 Дэвид В.Лестер               9.1
 Ю.Осетинская                 9.0
Э Мультимедиа Еуроп Филм      8.8
 Мокер Продакшн               8.8
 Куин Филмз Презент           8.8
 Э Мэфилм Дайлог Студио       8.8
 Карен Гилкрайст              8.8
В.Меньшов                     8.8
Джон Фавро                    8.8

### Результат не очень отражает суть, победители сняли лишь один фильм(

In [206]:
w_df[(w_df['producer'].notna()) & (w_df['producer'].str.contains('Ники Марвин'))].T

puNumber                                                       121029519
title                                                  Побег из Шоушенка
show_start_date                                               2019-10-18
type                                                      Художественный
film_studio                Уорнер Броз. Пикчерз, Кастл Рок Интертейнмент
production_country                                                   США
director                                                  Фрэнк Дарабонт
producer                         Лиз Глоцер, Дэвид В.Лестер, Ники Марвин
age_restriction                          «16+» - для детей старше 16 лет
refundable_support                                                   NaN
nonrefundable_support                                                NaN
budget                                                               NaN
financing_source                                                     NaN
ratings                                                              9.1
genres                                                             драма
box_office                                                           NaN
title_true                                             Побег из Шоушенка
first_lv_desc                                                       None
second_lv_desc                                                      None
alter_title                                                          NaN
title_true_len                                                        17
show_start_year                                                     2019
show_start_month                                                      10
age_limit                                                             16
counts_film_studio                                                   1.0
counts_production_country                                            0.0
counts_director                                                      0.0
counts_producer                                                      2.0
box_office_mean                                                      NaN
box_office_median                                                    NaN

# Задачи от бизнеса

1. Какой жанр имеет наибольший рейтинг? Учитывайте все жанры, к которым относится фильм. (задача на посчитать)

In [208]:
w_df_ganres = w_df[~w_df['genres'].isna()]

In [211]:
w_df_ganres['genres_list'] = w_df_ganres['genres'].str.split(',')

<ipython-input-211-3361362ab258>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  w_df_ganres['genres_list'] = w_df_ganres['genres'].str.split(',')


In [212]:
w_df_ganres_explode = w_df_ganres.explode('genres_list').drop_duplicates()
w_df_ganres_explode = w_df_ganres_explode.reset_index()
w_df_ganres_explode.shape

(15211, 31)

In [214]:
top_ganres = pd.pivot_table(w_df_ganres_explode, values='ratings',
                                                                                     index=['genres_list'],
                    aggfunc={'ratings': np.mean})


top_ganres.sort_values(by='ratings', ascending=False).head(10)

ratings
genres_list              
новости          7.800000
аниме            7.443836
документальный   7.202632
концерт          7.165000
короткометражка  7.066038
биография        7.040968
мюзикл           6.993636
спорт            6.961458
вестерн          6.906250
военный          6.905584

In [220]:
w_df[(w_df['genres'].notna()) & (w_df['genres'].str.contains('новости'))].T

puNumber                                                           126003515
title                                                     Реальная цена моды
show_start_date                                                   2015-11-25
type                                                                  Прочие
film_studio                Антолд Креатив, Лайф Ис Май Муви Интертейнмент...
production_country         Великобритания - Франция - Италия - Индия - Да...
director                                                        Эндрю Морган
producer                   Винсент Витторио, Ливия Джуджолли, Майкл Росс,...
age_restriction                              «16+» - для детей старше 16 лет
refundable_support                                                       NaN
nonrefundable_support                                                    NaN
budget                                                                   NaN
financing_source                                                         NaN
ratings                                                                  7.8
genres                                          документальный,новости,драма
box_office                                                          371909.0
title_true                                                Реальная цена моды
first_lv_desc                                                           None
second_lv_desc                                                          None
alter_title                                                              NaN
title_true_len                                                            18
show_start_year                                                         2015
show_start_month                                                          11
age_limit                                                                 16
counts_film_studio                                                       1.0
counts_production_country                                                9.0
counts_director                                                          0.0
counts_producer                                                          4.0
box_office_mean                                                     371909.0
box_office_median                                                   371909.0

### <font color='green'>выделил задачки с которыми хочу поработать</font>

2. Велик ли объем господдержки в фильмах? (надо посчитать долю для каждого фильма и обосновать много это или мало)

3. Какие фильмы лучше - с господдержкой или без? (надо предложить и обосновать понятие "лучше")

4. Составьте список успешных режиссеров. (предложить и обосновать понятие успешности)

### <font color='green'>5. Каких продюсеров можно назвать "восходящими звездами кинематографа?" (предложить показатель успешности и проанализовать динамику этого показателя)</font>

### <font color='green'>6. Влияет ли возрастное ограничение на успех фильма?</font>

### <font color='green'>7. Решите проблемы с пропусками в данных с помощью внешних источников.</font>

### <font color='green'>8. Сейчас наши коллеги-аналитики работают с данными в Экселе и тратят много времени на обработку. Данные проекта - результат из работы. Ваша задача указать проблемы в данных, которые они упустили. И показать преимущества pandas для подобных задач.</font>

### <font color='green'>9. Постройте модель предсказания будущего рейтинга фильма или успешности в финансовых показателях.</font>


## Пока не до конца успел сформулировать свои дальнейшие шаги, в общих чертах:

1. хочу разместить БД по фильмам в Google Cloude;

2. в БД накапливать/парсить данные - использовать Silenium;

3. там же собрать модель для оценки успешных продюссеров и продюссерских коллективов.